### PR TITLE
Fix reactive capability curve import

### PR DESCRIPTION
**Pull request by geofjamg**
Signed-off-by: Geoffroy Jamgotchian <geoffroy.jamgotchian@rte-france.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
#746 
#747 

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Bug fix

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java [hunk 1]
```diff
@@ -32,49 +36,89 @@ protected void convertReactiveLimits(Generator g) {
...
-                maxQ = minMaxQ[1];
-                rcca.beginPoint()
-                    .setP(p)
-                    .setMinQ(minQ)
-                    .setMaxQ(maxQ)
-                    .endPoint();
-            });
-            rcca.add();
+            } else {
+                missing("Empty curve");
```
Comment (MioRtia) [inline cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java]: In this case, we should still try to read `minQ` and `maxQ` in the traditional way.

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java [hunk 2]
```diff
@@ -32,49 +36,89 @@ protected void convertReactiveLimits(Generator g) {
...
+                PropertyBag d = curveData.get(0);
                 double p = d.asDouble("xvalue");
                 double minQ = d.asDouble("y1value");
                 double maxQ = d.asDouble("y2value");
-                // Check the point is valid
-                if (Double.isNaN(p) || Double.isNaN(minQ) || Double.isNaN(maxQ)) {
-                    String reason = String.format("Incomplete point p, minQ, maxQ = %f, %f, %f", p, minQ, maxQ);
-                    ignored("ReactiveCapabilityCurvePoint", reason);
-                    return;
+                if (checkPointValidity(p, minQ, maxQ)) {
```
Comment (MioRtia) [inline cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java]: if this point is not valid, we should still try to read `minQ` and `maxQ` in the traditional way.

**Comment by geofjamg** [comment]
> I agree with your modifications but in general, if no valid reactive capability curve is found, fixed reactive limits should be read to avoid cases where there is no reactive limits (which is bad practice).

But reactive capability curve and min/max are mutually exclusive? I do not see why in case of badly defined capability curve (either empty or single point), something could be available in min/max attribute.
There is no real problem for me to let a generator with unlimited reactif range in case of severe data issue.

**Comment by MioRtia** [comment]
> > I agree with your modifications but in general, if no valid reactive capability curve is found, fixed reactive limits should be read to avoid cases where there is no reactive limits (which is bad practice).
> 
> But reactive capability curve and min/max are mutually exclusive? I do not see why in case of badly defined capability curve (either empty or single point), something could be available in min/max attribute.
> There is no real problem for me to let a generator with unlimited reactif range in case of severe data issue.

In CGMES, reactive capability curve and min/max are not explicitly mutually exclusive. Reactive capability curve must be checked first if it exists but it is not conclusive, minQ and maxQ might be correctly defined and therefore used.

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java [hunk 3]
```diff
@@ -32,49 +36,89 @@ protected void convertReactiveLimits(Generator g) {
...
+                            ignored("reactive capability curve point", String.format("another point with same p (%f) and a disconnected reactive range", p));
+                        }
+                    } else {
+                        qRanges.put(p, qRange);
+                    }
+                });
+
+                // create curve
+                ReactiveCapabilityCurveAdder rcca = g.newReactiveCapabilityCurve();
+                for (Map.Entry<Double, Range<Double>> e : qRanges.entrySet()) {
```
Comment (MioRtia) [inline cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java]: What happens here if after checking duplicated points, `qRange` size is 1?

Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java]: Fixed in the new algo, check on point count is done after removing duplicate points.

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java (hunk 2 above)
Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java]: Fixed in the new algo, check on point count is done after removing duplicate points.

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java (hunk 1 above)
Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java]: Ok, done

File: cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java [hunk 4]
```diff
@@ -32,49 +36,93 @@ protected void convertReactiveLimits(Generator g) {
...
+                    double p = e.getKey();
+                    Range<Double> qRange = e.getValue();
+                    rcca.beginPoint()
+                            .setP(p)
+                            .setMinQ(qRange.lowerEndpoint())
+                            .setMaxQ(qRange.upperEndpoint())
+                            .endPoint();
+                }
+                rcca.add();
+            } else if (curveData.size() == 1) {
```
Comment (MioRtia) [inline cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java]: replace by `else if (qRanges.size() == 1)`

Comment (geofjamg) [reply cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/elements/AbstractReactiveLimitsOwnerConversion.java]: Fixed

**Pull request merged**